verifiers/internal/vsa/verifier.go

```diff
@@ -0,0 +1,215 @@
+	// 3. parse the VSA, verifying the predicateType.
+	vsa, err := extractSignedVSA(ctx, envelope, verificationOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	// 2. match the subject digests,
+	// 4. match the verifier ID,
+	// 5. match the expected resourceURI,
+	// 6. confirm the slsaResult is PASSED,
+	// 7. match the verifiedLevels,
+	// no other fields are checked.
+	err = matchExpectedValues(vsa, vsaOpts)
+	if err != nil {
+		return nil, err
+	}
+	vsaBytes, err := envelope.DecodeB64Payload()
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", serrors.ErrorInvalidDssePayload, err)
+	}
+	return vsaBytes, nil
+}
+
+// extractSignedVSA verifies the envelope signature and type and extracts the VSA from the envelope.
+func extractSignedVSA(ctx context.Context, envelope *dsse.Envelope, verificationOpts *options.VerificationOpts) (*vsa10.VSA, error) {
+	// 1. verify the envelope signature,
+	// 4. match the verfier with the public key: implicit because we accept a user-provided public key.
+	err := verifyEnvelopeSignature(ctx, envelope, verificationOpts)
+	if err != nil {
+		return nil, err
+	}
+	statement, err := utils.StatementFromEnvelope(envelope)
+	if err != nil {
+		return nil, err
+	}
+	// 3. parse the VSA, verifying the predicateType.
+	vsa, err := vsa10.VSAFromStatement(statement)
+	if err != nil {
+		return nil, err
+	}
+	return vsa, nil
+}
+
+// verifyEnvelopeSignature verifies the signature of the envelope.
+func verifyEnvelopeSignature(ctx context.Context, envelope *dsse.Envelope, verificationOpts *options.VerificationOpts) error {
+	signatureVerifier, err := sigstoreSignature.LoadVerifier(verificationOpts.PublicKey, verificationOpts.PublicKeyHashAlgo)
+	if err != nil {
+		return fmt.Errorf("%w: loading sigstore DSSE envolope verifier: %w", serrors.ErrorInvalidPublicKey, err)
+	}
+	envelopeVerifier, err := dsse.NewEnvelopeVerifier(&sigstoreDSSE.VerifierAdapter{
+		SignatureVerifier: signatureVerifier,
+		Pub:               verificationOpts.PublicKey,
+		PubKeyID:          *verificationOpts.PublicKeyID,
+	})
+	if err != nil {
+		return fmt.Errorf("%w: creating sigstore DSSE envelope verifier: %w", serrors.ErrorInvalidPublicKey, err)
+	}
+	_, err = envelopeVerifier.Verify(ctx, envelope)
+	if err != nil {
+		return fmt.Errorf("%w: verifying envelope: %w", serrors.ErrorNoValidSignature, err)
+	}
+	return nil
+}
+
+// matchExpectedValues checks if the expected values are present in the VSA.
+func matchExpectedValues(vsa *vsa10.VSA, vsaOpts *options.VSAOpts) error {
+	// 2. match the expected subject digests
+	if err := matchExepectedSubjectDigests(vsa, vsaOpts); err != nil {
+		return err
+	}
+	// 4. match the verifier ID
+	if err := matchVerifierID(vsa, vsaOpts); err != nil {
+		return err
+	}
+	// 5. match the expected resourceURI
+	if err := matchResourceURI(vsa, vsaOpts); err != nil {
+		return err
+	}
+	// 6. confirm the verificationResult is Passed
+	if err := confirmVerificationResult(vsa); err != nil {
```

Comment:
nit: be consustent with function names. Other functions are called `match*`, this one is called `confirm*`

Reply:
Only because, unlike the others, it doesn't compare the VSAs value with any user input.

Reply:
gotcha. just a nit. up to you :)